drivers/flash/spi_nor.c

```diff
@@ -540,6 +540,124 @@ static int spi_nor_wrsr(const struct device *dev,
 	return ret;
 }
 
+#if DT_INST_NODE_HAS_PROP(0, mxicy_lh_switch)
+
+/**
+ * @brief Read the configuration register.
+ *
+ * @note The device must be externally acquired before invoking this
+ * function.
+ *
+ * @param dev Device struct
+ *
+ * @return the non-negative value of the configuration register, or an error code.
+ */
+static int mxicy_rdcr(const struct device *dev)
+{
+	uint16_t reg;
+	enum { CMD_RDCR = 0x15 };
+	int ret = spi_nor_cmd_read(dev, CMD_RDCR, &reg, sizeof(reg));
+
+	if (ret == 0) {
+		ret = reg;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Write the configuration register.
+ *
+ * @note The device must be externally acquired before invoking this
+ * function.
+ *
+ * @param dev Device struct
+ * @param cr  The new value of the configuration register
+ *
+ * @return 0 on success or a negative error code.
+ */
+static int mxicy_wrcr(const struct device *dev,
+			uint16_t cr)
+{
+	/* The configuration register bytes on the Macronix MX25R devices are
+	 * written using the Write Status Register command where the configuration
+	 * register bytes are written as two extra bytes after the status register.
+	 * First read out the current status register to preserve the value.
+	 */
+	int sr = spi_nor_rdsr(dev);
+
+	if (sr < 0) {
+		LOG_ERR("Read status register failed: %d", sr);
+		return sr;
+	}
+
+	int ret = spi_nor_cmd_write(dev, SPI_NOR_CMD_WREN);
+
+	if (ret == 0) {
+		uint8_t data[] = {
+			sr,
+			cr & 0xFF,	/* Configuration register 1 */
+			cr >> 8		/* Configuration register 2 */
+		};
+
+		ret = spi_nor_access(dev, SPI_NOR_CMD_WRSR, NOR_ACCESS_WRITE, 0, data,
+			sizeof(data));
+		spi_nor_wait_until_ready(dev);
+	}
+
+	return ret;
+}
+
+static int mxicy_configure(const struct device *dev, uint8_t *jedec_id)
+{
+	/* Low-power/high perf mode is second bit in configuration register 2 */
+	enum { LH_SWITCH_BIT = 9 };
+	const uint8_t JEDEC_MACRONIX_ID = 0xc2;
+	const uint8_t JEDEC_MX25R_TYPE_ID = 0x28;
+	int current_cr, new_cr, ret;
+	/* lh_switch enum index:
+	 * 0: Ultra low power
+	 * 1: High performance mode
+	 */
+	const bool use_high_perf = DT_INST_ENUM_IDX(0, mxicy_lh_switch);
+
+	/* Only supported on Macronix MX25R Ultra Low Power series. */
+	if (jedec_id[0] != JEDEC_MACRONIX_ID || jedec_id[1] != JEDEC_MX25R_TYPE_ID) {
+		LOG_WRN("L/H switch not supported for device id: %02x %02x %02x", jedec_id[0],
+			jedec_id[1], jedec_id[2]);
+		return 0;
```

Comment:
> Shouldn't an error be returned in this case?

I was against returning error here, because I do not find it useful to error out when device does not support a feature, as the device will still function. We already log warning here that a configuration options has been set to device that does not seem to support it, no need to fail device configuration completely.
In the future we may add enums like "low-power-only" (or something like this), that may state that the option is required for the device, and in such case the configuration should fail.

This should allow scenarios I have described here https://github.com/zephyrproject-rtos/zephyr/pull/51659#issuecomment-1308473211function, when you can support different configuration options for different vendors, without knowing which device exactly is soldered in on that SPI.

Reply:
Okay. But then a comment here should mention that it is intentional. In the current form, it is a bit confusing.